### PR TITLE
[pickers] Add support for timezones on the adapters

### PR DESCRIPTION
**Pull request by flaviendelangle**
Extracted from #8261
Part of #8484

This PR focuses on the adapters and should not include any behavior changes on the pickers themselves.

- [x] Add 3 new methods `dateWithTimezone` (will replace `date` in v7), `getTimezone` and `setTimezone`
- [x] Add tests

### CODE REVIEW

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts [hunk 1]
```diff
@@ -262,47 +402,67 @@ export class AdapterDayjs implements MuiPickersAdapter<Dayjs, string> {
...
+      return value.isBefore(comparing, 'day');
+    }
+
+    // @ts-ignore
+    return !this.isSameDay(value, comparing) && value.utc() < comparing.utc();
   };
 
   public isWithinRange = (value: Dayjs, [start, end]: [Dayjs, Dayjs]) => {
-    return value.isBetween(start, end, null, '[]');
+    return value >= start && value <= end;
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts:453]: We don't need the power of `isBefore` and `isAfter` because we are not passing any unit.
The main advantage of this change is that `isBetween` / `isAfter` are very slow on dayjs with the `timezone` plugin, so I'm trying to use them as little as possible

File: packages/x-date-pickers/src/AdapterLuxon/AdapterLuxon.ts [hunk 2]
```diff
@@ -259,36 +296,42 @@ export class AdapterLuxon implements MuiPickersAdapter<DateTime, string> {
       return false;
     }
 
-    return this.date(value)!.equals(this.date(comparing)!);
+    return +this.date(value)! === +this.date(comparing)!;
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/AdapterLuxon/AdapterLuxon.ts:299]: Technically speaking, this is a breaking change for the Luxon adapter, because now `isEqual` returns true if you have 2 dates with the same timestamp but a different zone.

I think we can consider it to be a bug fix because the other adapters are not working the same way and are already based on the timestamp.

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts [hunk 3]
```diff
@@ -164,122 +277,343 @@ export const testCalculations: DescribeGregorianAdapterTestSuite = ({ adapter })
...
+      ).to.equal(false);
+    });
+
+    it('should work with different timezones', function test() {
+      if (!adapter.isTimezoneCompatible) {
+        this.skip();
+      }
+
+      // Both dates below have the same timestamp, but they are not in the same year when represented in their respective timezone.
+      // The adapter should still consider that they are in the same year.
```
Comment (alexfauquette) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:323]: This comment is somewhat disturbing. Would it be true to say that the `isSameYear` relies on the timezone of the first/second date provided?

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts [hunk 4]
```diff
@@ -262,47 +394,63 @@ export class AdapterDayjs implements MuiPickersAdapter<Dayjs, string> {
...
+    return this.isSame(value, comparing, 'YYYY-MM-DD');
   };
 
   public isSameHour = (value: Dayjs, comparing: Dayjs) => {
     return value.isSame(comparing, 'hour');
   };
 
   public isAfter = (value: Dayjs, comparing: Dayjs) => {
-    return value.isAfter(comparing);
+    return value > comparing;
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.ts:413]: `isAfter` just checks that A is after B, so we just have to compare there timestamp.

`isAfterYear` checks that the year of A is after the year of B.
So for that one, you have the question of "the year in which timezone ?"
The behavior I'm applying is to do the comparison in the timezone of the 1st param (`value`).
We can discuss if the other way around makes more sense (and adapt the param on this method call when adding timezone support to the pickers).

File: packages/x-date-pickers/src/AdapterLuxon/AdapterLuxon.test.tsx [hunk 5]
```diff
@@ -63,7 +68,7 @@ describe('<AdapterLuxon />', () => {
       ) => {
         const date = adapter.date('2020-02-01T23:44:00.000Z')!;
 
-        expect(adapter.format(date, format)).to.equal(expectedWithEn);
+        expect(cleanText(adapter.format(date, format))).to.equal(expectedWithEn);
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/x-date-pickers/src/AdapterLuxon/AdapterLuxon.test.tsx:71]: Yes, 
I can't find the source again, but the reason of my strange space is that the latest version of Chrome has a small bug that add this space in the `Intl` API instead of a regular space.

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts (hunk 3 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:323]: >  Would it be true to say that the isSameYear relies on the timezone of the first/second date provided?

Yes, all the `isSameXXX` `isAfterXXX` and `isBeforeXXX` are applied on the timezone of the 1st param.
`isSameYear` answers the question "for any A and B, is A and B in the same year when displayed in the timezone of A"

**Review by alexfauquette — APPROVED** [review]
It's good for me.

I don't have a strong opinion is `isAfterDay` should consider the TZ of the first or second argument. But it could be precise in the method comment:

```diff
- Check if the day of the reference date is after the day of the second date.
+ Check if the timestamp of the reference date is after the day of the second date (using the TZ of the second date).
```

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts [hunk 6]
```diff
@@ -164,122 +277,343 @@ export const testCalculations: DescribeGregorianAdapterTestSuite = ({ adapter })
...
+      expect(adapter.isEqual(testDateLocale, adapter.date(TEST_DATE_LOCALE_STRING))).to.equal(true);
+      expect(adapter.isEqual(null, testDateLocale)).to.equal(false);
+    });
+
+    it('should work with different timezones', function test() {
+      if (!adapter.isTimezoneCompatible) {
+        this.skip();
+      }
+
+      const londonTimezone = adapterTZ.setTimezone(testDateIso, 'Europe/London');
```
Comment (LukasTy) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:294]: Nitpick: This reads a bit ambiguously. 🙈 
By reading this line, I'd expect the `londonTimezone` to be just that, some sort of meta information about this particular timezone. 🤷 
What do you think about changing the method name to `setDateTimezone`? 🤔 
Also, update the `londonTimezone` to `dateWithLondonTZ` or `dateWithLondonTimezone`.

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts [hunk 7]
```diff
@@ -164,122 +277,343 @@ export const testCalculations: DescribeGregorianAdapterTestSuite = ({ adapter })
...
+
+      expect(adapterTZ.isAfterDay(londonTimezone, parisTimezone)).to.equal(false);
+      expect(adapterTZ.isAfterDay(parisTimezone, londonTimezone)).to.equal(false);
 
-    const nextDayLocale = adapter.addDays(testDateLocale, 1);
-    expect(adapter.isAfterDay(nextDayLocale, testDateLocale)).to.equal(true);
-    expect(adapter.isAfterDay(testDateLocale, nextDayLocale)).to.equal(false);
+      // Both dates below have the same day when represented in their respective timezone,
+      // But not when represented in the same timezone
+      // The adapter should consider that they are not in the same day
```
Comment (LukasTy) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:508]: Are they? 🤔 
It looks like you are adding `-1` hour to one of the value 🤔

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts [hunk 8]
```diff
@@ -164,122 +277,343 @@ export const testCalculations: DescribeGregorianAdapterTestSuite = ({ adapter })
...
+
+      expect(adapterTZ.isBeforeDay(londonTimezone, parisTimezone)).to.equal(false);
+      expect(adapterTZ.isBeforeDay(parisTimezone, londonTimezone)).to.equal(false);
 
-    const nextDayLocale = adapter.addDays(testDateLocale, -1);
-    expect(adapter.isBeforeDay(nextDayLocale, testDateLocale)).to.equal(true);
-    expect(adapter.isBeforeDay(testDateLocale, nextDayLocale)).to.equal(false);
+      // Both dates below have the same day when represented in their respective timezone,
+      // But not when represented in the same timezone
+      // The adapter should consider that they are not in the same day
```
Comment (LukasTy) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:605]: Same question regarding the correctness of this comment, because there is a manipulation of hours happening. 🤔

**Review by LukasTy — APPROVED** [review]
Massive and insanely important greatly executed feature! 🚀 💯 

Some minor nitpicks and a general question regarding the variable name correctness/clarity. 😉

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts (hunk 7 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:508]: Could you elaborate?
For me the comment is correct.
When represented in their respective timezone, the dates are on the same day.
When represented in the timezone of the 1st date, the dates are not on the same day (because of the 1 hour gap).

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts (hunk 6 above)
Comment (LukasTy) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:294]: Hm... I double-checked the existing adapter methods and it looks like all of them omit the `Date` part in the method name. 🤔 
Maybe sticking to the same convention is the best approach..? 🤔 
I would maybe rephrase my comment to clarify that the variable name confused me the most. 🙈 
What do you think about only renaming the variables to better express what kind of value they hold?

File: packages/x-date-pickers/src/models/timezone.ts [hunk 9]
```diff
@@ -0,0 +1 @@
+export type PickerTimezone = 'default' | 'system' | 'UTC' | string;
```
[the remark this answers is not in the document]
Comment (LukasTy) [reply packages/x-date-pickers/src/models/timezone.ts]: > I used singular for the field types (FieldSectionType, FieldSectionContentType, FieldValueType`), maybe we should consider renaming those on v7.

I'm not sure that it makes a big difference. 🤔 What caught my eye was that this new type is singular, while others starting with `Picker` are plural. If the `Field` types are consistent between themselves, that's good enough IMHO. 🤔

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts (hunk 6 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:294]: `dateInLondonTZ` would work for you ?

Comment (LukasTy) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:294]: Sounds great 👌 🙈

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts (hunk 7 above)
Comment (LukasTy) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:508]: Nevermind. Looked at it with fresh eyes and it is correct. Sorry for the confusion 🙈 😊

Comment (LukasTy) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:508]: I can't think of a clear improvement to the comment ATM.
It's a complex topic anyways like you mentioned. 🙈

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts [hunk 10]
```diff
@@ -41,6 +39,121 @@ export const testCalculations: DescribeGregorianAdapterTestSuite = ({ adapter })
...
+        };
+
+        testTodayZone('system');
+        testTodayZone('UTC');
+        testTodayZone('America/New_York');
+      }
+    });
+  });
+
+  it('Method: getTimezone', () => {});
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:132]: Indeed, I added some tests
